pipeline_dp/budget_accounting.py

```diff
@@ -8,139 +9,165 @@
-    of the corresponding BudgetAccount is called.
+    NoiseKind defines the kind of noise distribution.
+    _noise_standard_deviation is the minimized noise standard deviation.
+    (_eps, _delta) are parameters of (eps, delta)-differential privacy
     """
+    noise_kind: NoiseKind
+    _noise_standard_deviation: float = None
     _eps: float = None
     _delta: float = None
 
     @property
-    def eps(self):
-        """Parameter of (eps, delta)-differential privacy.
+    def noise_standard_deviation(self):
+        """Noise value for the mechanism.
 
         Raises:
-            AssertionError: The privacy budget is not calculated yet.
+            AssertionError: The noise value is not calculated yet.
         """
+        if self._noise_standard_deviation is None:
+            raise AssertionError(
+                "Noise standard deviation is not calculated yet.")
+        return self._noise_standard_deviation
+
+    @property
+    def eps(self):
+        """Parameter of (eps, delta)-differential privacy.
+               Raises:
+                   AssertionError: The privacy budget is not calculated yet.
+       """
         if self._eps is None:
             raise AssertionError("Privacy budget is not calculated yet.")
         return self._eps
 
     @property
     def delta(self):
         """Parameter of (eps, delta)-differential privacy.
-
-        Raises:
-            AssertionError: The privacy budget is not calculated yet.
+                Raises:
+                    AssertionError: The privacy budget is not calculated yet.
         """
         if self._delta is None:
             raise AssertionError("Privacy budget is not calculated yet.")
         return self._delta
 
-    def set_eps_delta(self, eps, delta):
+    def set_eps_delta(self, eps: float, delta: float) -> None:
+        """Set parameters for (eps, delta)-differential privacy.
+
+        delta can be None.
+        Raises:
+            AssertionError: eps must not be None.
+        """
+        if eps is None:
+            raise AssertionError("eps must not be None.")
         self._eps = eps
         self._delta = delta
+        return
+
+    def use_delta(self) -> bool:
+        return self.noise_kind == NoiseKind.GAUSSIAN
 
 
 @dataclass
-class RequestedBudget:
-    """Manages the budget requested for an operation."""
-    budget: Budget
+class MechanismSpecInternal:
+    """Stores sensitivity and weight not exposed in MechanismSpec."""
+    sensitivity: float
     weight: float
-    use_eps: bool
-    use_delta: bool
+    mechanism_spec: MechanismSpec
 
 
-class BudgetAccountant:
+class BudgetAccountant(abc.ABC):
+    """Base class for budget accountants."""
+
+    @abc.abstractmethod
+    def request_budget(self,
+                       noise_kind: NoiseKind,
+                       sensitivity: float = 1,
+                       weight: float = 1) -> MechanismSpec:
+        pass
+
+    @abc.abstractmethod
+    def request_budget(self):
+        pass
```

Comment:
Please add abstract method `compute_budgets()`

Reply:
Sorry about this I accidentally doubled this method.